src/datarobotx/idp/custom_application_source.py

```diff
@@ -0,0 +1,320 @@
+                    "label",
+                    f"{kwargs.get('label', '')}{' - ' if kwargs.get('label') else ''}[{version_token}]",
+                )
+            )
+
+            encoder = MultipartEncoder(fields=upload_data)
+            headers = {"Content-Type": encoder.content_type}
+
+            if from_previous:
+                response = client.request(
+                    method="PATCH",
+                    url=f"customApplicationSources/{custom_application_source_id}/versions/{previous_custom_application_source_version_id}/",
+                    data=encoder,
+                    headers=headers,
+                )
+            else:
+                response = client.request(
+                    method="POST",
+                    url=f"customApplicationSources/{custom_application_source_id}/versions/",
+                    data=encoder,
+                    headers=headers,
+                )
+            new_version = response.json()["id"]
+
+        return str(new_version)
+
+
+# Public functions
+def get_or_create_custom_application_source_version(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+    **kwargs: Any,
+) -> str:
+    """Get or create a custom application source version.
+
+    Parameters
+    ----------
+    endpoint : str
+        The DataRobot endpoint.
+    token : str
+        The DataRobot API token.
+    custom_application_source_id : str
+        runtime_parameter_values : Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]]
+    kwargs : Any
+
+    Returns
+    -------
+    str
+        The ID of the custom application source version.
+    """
+    return _get_or_create_custom_application_source_version(
+        from_previous=False,
+        endpoint=endpoint,
+        token=token,
+        custom_application_source_id=custom_application_source_id,
+        runtime_parameter_values=runtime_parameter_values,
+        **kwargs,
+    )
+
+
+def get_or_create_custom_application_source_version_from_previous(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    previous_custom_application_source_version_id: str,
+    runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+    **kwargs: Any,
+) -> str:
+    """Get or create a custom application source version from a previous version.
+
+    Parameters
+    ----------
+    endpoint : str
+        The DataRobot endpoint.
+    token : str
+        The DataRobot API token.
+    custom_application_source_id : str
+    previous_custom_application_source_version_id : str
+    runtime_parameter_values : Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]]
+    kwargs : Any
+
+    Returns
+    -------
+    str
+        The ID of the custom application source version.
+    """
+    return _get_or_create_custom_application_source_version(
```

Comment:
Ok I get the idea now. Small nit but consider doing: 
1. Only one public function `get_or_create_custom_application_source_version`
2. `previous_custom_application_source_version_id` is an optional argument which defaults to None
3.  If a `previous_custom_application_source_version_id` is supplied, we assume a patch request in the helper. Otherwise post request
4. `from_previous` is removed from helper function

Reply:
I had that, but then decided to follow the custom_model_version API here. Where dr SDK also has two methods, and we have two different functions

Reply:
I just found out that from_previous works differently to custom models. Custom models you get a new version with the content freom before. Here you get the same version with new content (rather interestingly..)

I think the function needs to be renamed

Reply:
Or maybe the function should be removed from idp?

Reply:
yeah. thank you for coming to my ted talk. I agree with you